File: algorithm_library/python/data-structure/TwoWayList.py
# 双方向連結リスト https://atcoder.jp/contests/abc344/editorial/9487
# 挿入・削除 O(1)
# 要素の重複不可


class TwoWayList:
    def __init__(self, L):
        self.head = -1
        self.tail = pow(10, 18)
        self.L = [self.head] + L + [self.tail]
        self.front, self.back = dict(), dict()
        for i in range(len(self.L) - 1):
            self.back[self.L[i]] = self.L[i + 1]  # xの後ろ
            self.front[self.L[i + 1]] = self.L[i]  # xの前

    def insert_front(self, x, y):  # xの前にyを挿入する
        # 現在 a -> x
        # a -> y -> x
        a = self.front[x]
        self.back[a] = y
        self.front[x] = y
        self.back[y] = x
        self.front[y] = a

    def insert_back(self, x, y):  # xの直後にyを挿入する
        # 現在 x -> b
        # x -> y -> b
        b = self.back[x]
        self.back[x] = y
        self.front[y] = x
        self.back[y] = b
        self.front[b] = y

    def del_x(self, x):  # xを削除する
        a = self.front[x]
        b = self.back[x]
        self.back[a] = b
        self.front[b] = a
        del self.front[x]
        del self.back[x]

    def res_LIST(self):  # リストを返す
        ans = list()
        crr = self.back[self.head]
        while crr != self.tail:
            ans.append(crr)
            crr = self.back[crr]
        return ans

File: algorithm_library/cpp-from-py/data-structure/TwoWayList.cpp
#include <bits/stdc++.h>
using namespace std;

struct TwoWayList {
    static constexpr long long HEAD = -1;
    static constexpr long long TAIL = (long long)1e18;
    unordered_map<long long, long long> front, back;

    TwoWayList(const vector<long long>& L) {
        vector<long long> v;
        v.push_back(HEAD);
        v.insert(v.end(), L.begin(), L.end());
        v.push_back(TAIL);
        for (int i = 0; i + 1 < (int)v.size(); i++) {
            back[v[i]] = v[i + 1];
            front[v[i + 1]] = v[i];
        }
    }

    void insert_front(long long x, long long y) {
        long long a = front[x];
        back[a] = y;
        front[x] = y;
        back[y] = x;
        front[y] = a;
    }

    void insert_back(long long x, long long y) {
        long long b = back[x];
        back[x] = y;
        front[y] = x;
        back[y] = b;
        front[b] = y;
    }

    void del_x(long long x) {
        long long a = front[x], b = back[x];
        back[a] = b;
        front[b] = a;
        front.erase(x);
        back.erase(x);
    }

    vector<long long> res_LIST() const {
        vector<long long> ans;
        long long cur = back.at(HEAD);
        while (cur != TAIL) {
            ans.push_back(cur);
            cur = back.at(cur);
        }
        return ans;
    }
};
